File: src/app/canvas.ts
/* ------------------------------------------------------------------------ */
/* Canvas */
/* ------------------------------------------------------------------------ */

function draw() {
    var currentNote = last16thNoteDrawn;
    var currentTime = audioContext.currentTime;

    while (notesInQueue.length && notesInQueue[0].time < currentTime) {
        currentNote = notesInQueue[0].note;
        notesInQueue.splice(0, 1);   // remove note from queue
    }

    // We only need to draw if the note has moved.
    if (last16thNoteDrawn != currentNote) {
        var x = Math.floor(canvas.width / 18);
        canvasContext.clearRect(0, 0, canvas.width, canvas.height);
        for (var i = 0; i < 16; i++) {
            canvasContext.fillStyle = (currentNote == i) ?
                ((currentNote % 4 === 0) ? "red" : "blue") : "black";
            canvasContext.fillRect(x * (i + 1), x, x / 2, x / 2);
        }
        last16thNoteDrawn = currentNote;
    }

    // set up to draw again
    requestAnimFrame(draw);
}

window.requestAnimFrame = (function () {
    return window.requestAnimationFrame ||
        window.webkitRequestAnimationFrame ||
        window.mozRequestAnimationFrame ||
        window.oRequestAnimationFrame ||
        window.msRequestAnimationFrame ||
        function (callback) {
            window.setTimeout(callback, 1000 / 60);
        };
})();

function resetCanvas(e) {
    // resize the canvas - but remember - this clears the canvas too.
    canvas.width = window.innerWidth;
    canvas.height = window.innerHeight;

    //make sure we scroll to the top left.
    window.scrollTo(0, 0);
}

function initCanvas() {
    const container = document.createElement('div');
    container.className = "container";
    this.canvas = document.createElement('canvas');
    this.canvasContext = canvas.getContext('2d');
    this.canvas.width = window.innerWidth;
    this.canvas.height = window.innerHeight;
    document.body.appendChild(container);
    container.appendChild(canvas);
    this.canvasContext.strokeStyle = "#ffffff";
    this.canvasContext.lineWidth = 2;

    window.onorientationchange = this.resetCanvas;
    window.onresize = this.resetCanvas;

    requestAnimFrame(draw);
}

File: bs_tea_simple/src/canvas.ml

type _window = <
  width : int;
  height : int;
  innerHeight : int;
  innerWidth : int
> Js.t

type canvas = {
  mutable height : int;
  mutable width: int
} [@@bs.deriving]

type context = {
  keep: int;
} [@@bs.deriving]

external window : _window = "window" [@@bs.val]
external get_canvas : string -> canvas = "getElementById" [@@bs.val] [@@bs.scope "document"]
external get_context : canvas -> string -> context = "getContext" [@@bs.send]
external fill_style_set : context -> string -> unit = "fillStyle" [@@bs.set]
external clear_rect : context -> int -> int -> int -> int -> unit = "clearRect" [@@bs.send]
external fill_rect : context -> int -> int -> int -> int -> unit = "fillRect" [@@bs.send]
external fill_text : context -> string -> int -> int -> unit = "fillText" [@@bs.send]
external font_set : context -> string -> unit = "font" [@@bs.set]
external scroll_to : int -> int -> unit = "scrollTo" [@@bs.val] [@@bs.scope "window"]
external request_animation_frame : (unit -> unit) -> unit = "requestAnimationFrame" [@@bs.val] [@@bs.scope "window"]


class t canvas_id =
  let c = get_canvas canvas_id in
  object
    val mutable canvas = c
    val mutable canvas_context = get_context c "2d"
    val mutable current_note = 0
    val mutable current_time = 0
    val note_buffer = Vector.make 16 ~dummy:0

    method update ~time ~note =
      current_time <- time;
      current_note <- note;
      while (not (Vector.is_empty note_buffer)) && (Vector.get note_buffer 0) == 0 do
        current_note <- Vector.pop note_buffer;
      done;


    method draw () =
      ()

    method reset_canvas () =
      canvas.width <- window##innerWidth;
      canvas.height <- window##innerHeight;
      scroll_to 0 0
  end
